File: CSharp/Engine/Runtime/Core/System/Timer.cs



#if PLATFORM_WINDOWS
using static Engine.Windows.APISet.ProfileAPI;
#endif


namespace Engine
{
	public class Timer
	{
		/// <summary>
		/// Time point
		/// </summary>
		/// <remarks> Unit is second </remarks>
		private float start;


		public Timer() => Reset();


		/// <summary>
		/// Record current system time
		/// </summary>
		public void Reset() => start = CurrentSystemTime();

		/// <summary>
		/// Time elapsed in seconds
		/// </summary>
		public float Elapsed() => CurrentSystemTime() - start;


		private static double period = -1;
		private static float CurrentSystemTime()
		{
#if PLATFORM_WINDOWS
			if( period < 0 )
			{
				LARGE_INTEGRE frequency = default;
				QueryPerformanceFrequency(ref frequency);

				period = 1.0 / frequency;
			}

			LARGE_INTEGRE counter = default;
			QueryPerformanceCounter(ref counter);

			return ( (float)( period * counter ) );
#else
#error "No implementation"
#endif
		}
	}
}

File: C/Engine/Source/Runtime/Core/System/Timer.c
#include "Timer.h"

#if defined(PLATFORM_WINDOWS)
#include "../../Platform/Windows/Windows.h"
#endif



static RealType CurrentSystemTime()
{
#if defined(PLATFORM_WINDOWS)
	static Float64 period = -1;
	if ( period < 0 )
	{
		LARGE_INTEGER frequency;
		QueryPerformanceFrequency(&frequency);

		period = 1.0 / frequency.QuadPart;
	}

	LARGE_INTEGER counter;
	QueryPerformanceCounter(&counter);

	return period * counter.QuadPart;
#else
#error "No implementation"
#endif
}


static Timer Construct()
{
	Timer ret;
	FTimer.Reset( &ret );

	return ret;
}

static Void Reset(Timer* instance)
{
	instance->start = CurrentSystemTime();
}

static RealType Elapsed(const Timer* instance)
{
	return CurrentSystemTime() - instance->start;
}


struct FTimer FTimer = 
{
	Construct,
	Reset,
	Elapsed,
};
